docs: add Universal Deployer Contract (UDC) section (#1011)
starknet-io/starknet-docs: components/Starknet/modules/architecture_and_concepts/pages/Smart_Contracts/universal-deployer.adoc

:deploy-syscall: xref:./system-calls-cairo1.adoc#deploy[deploy syscall]

[id="universal_deployer"]
= Universal Deployer Contract

The Universal Deployer Contract (UDC) is a singleton smart contract that wraps the {deploy-syscall} to expose it to any contract that doesn’t implement it, such as account contracts. You can think of it as a standardized generic factory for Starknet contracts.

> And since Starknet has no deployment transaction type, it offers a canonical way to deploy smart contracts by following the https://community.starknet.io/t/snip-deployer-contract-interface/2772[standard deployer interface] and emitting a `ContractDeployed` event. For more details on thought process and decision making, refer to the https://community.starknet.io/t/universal-deployer-contract-proposal/1864[Universal Deployer Contract Proposal].

Author: i guess every way to deploy is non-canonical. the udc is a standard that aims to be canonical by being deployed at a given address and act as a singleton factory contract. the intention is to have a single, canonical instance baked into smart contracts, sdks, tooling, etc.
for example, using the same instance of the UDC in different chains could enable reproducible origin-dependent deployments.

> And since Starknet has no deployment transaction type, it offers a canonical way to deploy smart contracts by following the https://community.starknet.io/t/snip-deployer-contract-interface/2772[standard deployer interface] and emitting a `ContractDeployed` event. For more details on thought process and decision making, refer to the https://community.starknet.io/t/universal-deployer-contract-proposal/1864[Universal Deployer Contract Proposal].

Reviewer: OK. I changed *canonical* to *standardized*.

> NOTE: The UDC address is `0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf` in mainnet, testnets, and starknet-devnet. This address may change in the future as it gets migrated to Cairo 1+.

Author: 0

> NOTE: The UDC address is `0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf` in mainnet, testnets, and starknet-devnet. This address may change in the future as it gets migrated to Cairo 1+.

Reviewer: I changed _Cairo 1+_ to _a modern version of Cairo_.

----
use openzeppelin::utils::universal_deployer::IUniversalDeployerDispatcher;

const UDC_ADDRESS = 0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf;
let udc = IUniversalDeployerDispatcher { contract_address: UDC_ADDRESS }

let class_hash = 0x5c478ee27f2112411f86f207605b2e2c58cdb647bac0df27f660ef2252359c6;
let salt = 1234567879;
let unique = false;
let mut calldata = ArrayTrait::<felt252>::new();

let deployed_addr = udc.deployContract(class_hash, salt, unique, calldata.span());
----

=== Deployment types

The Universal Deployer Contract offers two types of addresses to deploy: origin dependent and origin independent.
As the names suggest, the origin-dependent type includes the deployer's address in the address calculation;
whereas, the origin-independent type does not.
The `unique` boolean parameter ultimately determines the type of deployment.

CAUTION: When deploying a contract that uses `get_caller_address` in the constructor calldata, remember that the UDC (and not the account) deploys that contract.
Therefore, querying `get_caller_address` in a contract's constructor will return the UDC's address and *not* the account's address.

==== Origin dependent

> Making deployments dependent of the origin address, users can "reserve" themselves a whole address space from squatting by someone else.

Reviewer: 1. Who is the user here? Isn't this for developers?
2. Why is "reserve" in quotation marks? Is it a special term in this context? Quotation marks signifies sarcasm, so we don't use them except for specific cases that require them. So I'm not sure what you're trying to say.
3. It says "dependent of the origin address". Do you mean "dependent on the origin address" or "independent of the origin address"?

Author: 1. the user is whoever uses (calls) the contract. any account can directly call this contract, so you could call it from your wallet if you wanted to. it's not constrained to programatic use by developers.
2. we can drop the quotations marks
3. dependent on the origin address

Reviewer: OK. PTAL:

```
By making deployments dependent upon the origin address, users can reserve a whole address space to prevent someone else from taking ownership of the address.

Only the owner of the origin address can deploy to those addresses.
```

To deploy a unique contract address pass:

[,js]
----
let deployed_addr = udc.deployContract(class_hash, salt, true, calldata.span());
----

==== Origin independent

Origin-independent contract deployments create contract addresses independent of the deployer and the UDC instance.
Instead only the class hash, salt, and constructor arguments determine the address.
This type of deployment enables redeployments of accounts and known systems across multiple networks.
To deploy a reproducible deployment, set `unique` to `false`.

[,js]
----
let deployed_addr = udc.deployContract(class_hash, salt, false, calldata.span());
----

== Deploying the UDC

NOTE: The Universal Deployer Contract has already been deployed on most networks and development environments.
The standard requires the UDC to be deployed passing `deploy_from_zero=true` and `salt=0` as arguments to the {deploy-syscall}. This results in a deterministic and predictable address across all instances of Starknet, facilitating SDK integration and reproducibility of deployments.

== API specification

=== Methods

> ==== `deployContract`

Author: Do you think pasting the full implementation in the docs will help? it seemed enough to me with the interface and behavior explainer. The implementation can be found in block explorers with [the address](https://starknet-io.github.io/starknet-docs/pr-1011/documentation/architecture_and_concepts/Smart_Contracts/universal-deployer/#udc_address), and in the proposal linked at the beginning.

> ==== `deployContract`

Reviewer: @martriay I'm fine with keeping the interface in the docs. I just want to be able to link to where the function is defined, I trust you to know if the interface is enough or if a developer would need to see the full implementation.

What's important to provide in the doc is whatever they need to use the function: 
- function Description
- description of each parameter, including type
- similar description of each return value (or if none, then state None)
- events that are triggered (if none, then state None)
- event attributes
- state mutability
- links to the function/interface definition

[id="universal_deployer_contract"]
= Universal Deployer Contract (UDC)

:deploy-syscall: xref:architecture_and_concepts:Smart_Contracts/system-calls-cairo1.adoc#deploy[deploy syscall]

The Universal Deployer Contract (UDC) is a singleton smart contract that wraps the {deploy-syscall} to expose it to any contract that doesn’t implement it, such as account contracts. You can think of it as a standardized generic factory for Starknet contracts.

And since Starknet has no deployment transaction type, it offers a canonical way to deploy smart contracts by following the https://community.starknet.io/t/snip-deployer-contract-interface/2772[standard deployer interface] and emitting a `ContractDeployed` event.

> For more information see the https://community.starknet.io/t/snip-deployer-contract-interface/2772[proposal for the standard deployer interface].

Reviewer: @andrew-fleming Thanks for the suggestion, but unlike other writing styles, such as academic, or journalistic, or entertainment, one of the principles of technical writing is to be consistent: Use the same phrasing for the same type of information.

For details on the motivation and the decision making process, see the https://community.starknet.io/t/universal-deployer-contract-proposal/1864[Universal Deployer Contract proposal].

== UDC address

The UDC address is `0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf` in Mainnet, Goerli testnet, Sepolia testnet, and starknet-devnet. This address might change in the future when it is migrated to a modern version of Cairo.

== Interface

[source,cairo]
----
trait IUniversalDeployer {
    fn deployContract(
        class_hash: ClassHash,
        salt: felt252,
        unique: bool,
        calldata: Span<felt252>
    ) -> ContractAddress;
}
----

== Deploying a contract with the UDC

.Procedure

. Declare the contract with a `DECLARE` transaction, or ensure that the contract has been declared.
+
For more information, see the xref:architecture_and_concepts:Network_Architecture/transactions.adoc#declare-transaction[`DECLARE` transaction].
. Send a `deployContract` transaction to the UDC.

.Example implementation in Cairo:

[source,cairo]
----

> use openzeppelin::utils::universal_deployer::IUniversalDeployerDispatcher;

Reviewer 2: ```suggestion
use openzeppelin::universal_deployer::interface::IUniversalDeployerDispatcher;
```

Reviewer 2: I think this path is more likely to be consistent with our codebase.

const UDC_ADDRESS = 0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf;
let udc = IUniversalDeployerDispatcher { contract_address: UDC_ADDRESS }

let class_hash = 0x5c478ee27f2112411f86f207605b2e2c58cdb647bac0df27f660ef2252359c6;
let salt = 1234567879;

> let unique = false;

Reviewer 2: +1 to at least making the example compatible with the compiler

> let unique = false;

Reviewer: @andrew-fleming, @ericnordelo Can you propose a specific change?

> let unique = false;

Reviewer 2: ```
use starknet::{ClassHash, ContractAddress, class_hash_const};

#[starknet::interface]
trait IUniversalDeployer<TContractState> {
    fn deployContract(
        ref self: TContractState,
        class_hash: ClassHash,
        salt: felt252,
        unique: bool,
        calldata: Span<felt252>
    ) -> ContractAddress;
}

const UDC_ADDRESS: felt252 = 0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf;

fn deploy() {
    let dispatcher = IUniversalDeployerDispatcher {
        contract_address: UDC_ADDRESS.try_into().unwrap()
    };

    let class_hash = class_hash_const::<
        0x5c478ee27f2112411f86f207605b2e2c58cdb647bac0df27f660ef2252359c6
    >();
    let salt = 1234567879;
    let unique = false;
    let mut calldata = array![];

    let deployed_addr = dispatcher.deployContract(class_hash, salt, unique, calldata.span());
}
```

> let unique = false;

Reviewer 2: Something like this should work

> let unique = false;

Reviewer: I replaced the previous example beginning at line 44 with this one.

[id="universal_deployer_contract"]
= Universal Deployer Contract (UDC)

:deploy-syscall: xref:architecture_and_concepts:Smart_Contracts/system-calls-cairo1.adoc#deploy[deploy syscall]

The Universal Deployer Contract (UDC) is a singleton smart contract that wraps the {deploy-syscall} to expose it to any contract that doesn’t implement it, such as account contracts. You can think of it as a standardized generic factory for Starknet contracts.

> And since Starknet has no deployment transaction type, it offers a canonical way to deploy smart contracts by following the https://community.starknet.io/t/snip-deployer-contract-interface/2772[standard deployer interface] and emitting a `ContractDeployed` event.

Reviewer: ```suggestion
And since Starknet has no deployment transaction type, it offers a standardized way to deploy smart contracts by following the https://community.starknet.io/t/snip-deployer-contract-interface/2772[standard deployer interface] and emitting a `ContractDeployed` event.
```

For more information see the https://community.starknet.io/t/snip-deployer-contract-interface/2772[proposal for the standard deployer interface].
For details on the motivation and the decision making process, see the https://community.starknet.io/t/universal-deployer-contract-proposal/1864[Universal Deployer Contract proposal].

== UDC address

The UDC address is `0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf` in Mainnet, Goerli testnet, Sepolia testnet, and starknet-devnet. This address might change in the future when it is migrated to a modern version of Cairo.

== Interface

[source,cairo]
----
trait IUniversalDeployer {
    fn deployContract(
        class_hash: ClassHash,
        salt: felt252,
        unique: bool,
        calldata: Span<felt252>
    ) -> ContractAddress;
}
----

== Deploying a contract with the UDC

.Procedure

. Declare the contract with a `DECLARE` transaction, or ensure that the contract has been declared.
+
For more information, see the xref:architecture_and_concepts:Network_Architecture/transactions.adoc#declare-transaction[`DECLARE` transaction].

> . Send a `deployContract` transaction to the UDC.

Reviewer: ```suggestion
. Call the `deployContract` function in the UDC.
```

.Example implementation in Cairo:

[source,cairo]
----
#[starknet::interface]
trait IUniversalDeployer<TContractState> { 
    fn deployContract( 
        ref self: TContractState,
        class_hash: ClassHash,
        salt: felt252, 
        unique: bool,
        calldata: Span<felt252> 
    ) -> ContractAddress; } 
   
const UDC_ADDRESS: felt252 = 0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf; 
   
fn deploy() -> ContractAddress { 
    let dispatcher = IUniversalDeployerDispatcher { 
        contract_address: UDC_ADDRESS.try_into().unwrap() 
    };

    // deployment parameters
    let class_hash = class_hash_const::<
       0x5c478ee27f2112411f86f207605b2e2c58cdb647bac0df27f660ef2252359c6
    >();
    let salt = 1234567879; 
    let unique = false; 
    let mut calldata = array![];
        
    // the UDC returns the deployed contract address

>     dispatcher.deployContract(class_hash, salt, unique, calldata.span()); 

Reviewer: ```suggestion
    dispatcher.deployContract(class_hash, salt, unique, calldata.span())
```

}
----

== Deployment types

> The Universal Deployer Contract offers two types of addresses to deploy: origin dependent and origin independent.

Reviewer: ```suggestion
The Universal Deployer Contract offers two types of addresses to deploy: origin-dependent and origin-independent.
```

> As the names suggest, the origin-dependent type includes the deployer's address in the address calculation;

Reviewer: ```suggestion
As the names suggest, the origin-dependent type includes the deployer's address in the address calculation,
```

whereas, the origin-independent type does not.
The `unique` boolean parameter ultimately determines the type of deployment.

> [CAUTION]

Reviewer: ```suggestion
[IMPORTANT]
```

====

> When deploying a contract that uses `get_caller_address` in the constructor calldata, remember that the UDC (and not the account) deploys that contract.

Reviewer: ```suggestion
When deploying a contract that uses `get_caller_address` in the constructor calldata, remember that the UDC,  not the account, deploys that contract.
```

> Therefore, querying `get_caller_address` in a contract's constructor will return the UDC's address and *not* the account's address.

Reviewer: ```suggestion
Therefore, querying `get_caller_address` in a contract's constructor returns the UDC's address, _not the account's address_.
```

====

> === Origin dependent

Reviewer: ```suggestion
=== Origin-dependent
```

By making deployments dependent upon the origin address, users can reserve a whole address space to prevent someone else from taking ownership of the address.

Only the owner of the origin address can deploy to those addresses.

Achieving this type of deployment necessitates that the origin sets `unique` to `true` in the `deployContract` call.
Under the hood, the function call leverages the origin's address and creates a hashchain by hashing the origin's address with the given salt.

To deploy a unique contract address pass:

[,js]
----
let deployed_addr = udc.deployContract(class_hash, salt, true, calldata.span());
----

> === Origin independent

Reviewer: ```suggestion
=== Origin-independent
```
